File: Wheyt Watcher/BackupManager.swift
import Foundation
import SwiftData

enum BackupManager {

    // MARK: - Export

    static func buildPayload(
        profile: UserProfile?,
        foodEntries: [FoodLogEntry],
        trainings: [TrainingSession],
        weightLogs: [WeightLog],
        measurementLogs: [BodyMeasurementLog],
        dayStatuses: [DayStatus],
        snapshots: [DailyTargetSnapshot],
        favorites: [FavoriteFood],
        savedMeals: [SavedMeal],
        foodProducts: [FoodProduct],
        mealTemplates: [MealTemplate]
    ) -> BackupPayload {

        BackupPayload(
            exportedAt: Date(),

            profile: profile.map { p in
                ProfileDTO(
                    name: p.name,
                    age: p.age,
                    sex: p.sex.rawValue,
                    heightCm: p.heightCm,
                    currentWeightKg: p.currentWeightKg,
                    estimatedBodyFatPercentage: p.estimatedBodyFatPercentage,
                    goalMode: p.goalMode.rawValue,
                    goalPace: p.goalPace.rawValue,
                    activityLevel: p.activityLevel.rawValue,
                    createdAt: p.createdAt
                )
            },

            goalPeriods: (profile?.goalPeriods ?? []).map { period in
                GoalPeriodDTO(
                    startDate: period.startDate,
                    durationWeeks: period.durationWeeks,
                    goalMode: period.goalMode.rawValue,
                    goalPace: period.goalPace.rawValue,
                    isActive: period.isActive,
                    completedAt: period.completedAt,
                    calorieAdjustment: period.calorieAdjustment,
                    lastCheckInDate: period.lastCheckInDate
                )
            },

            foodEntries: foodEntries.map { entry in
                FoodLogEntryDTO(
                    date: entry.date,
                    mealCategory: entry.mealCategory.rawValue,
                    name: entry.name,
                    grams: entry.grams,
                    calories: entry.calories,
                    proteinGrams: entry.proteinGrams,
                    carbsGrams: entry.carbsGrams,
                    fatGrams: entry.fatGrams,
                    fiberGrams: entry.fiberGrams,
                    note: entry.note
                )
            },

            trainings: trainings.map { session in
                TrainingSessionDTO(
                    date: session.date,
                    type: session.type.rawValue,
                    durationMinutes: session.durationMinutes,
                    rpe: session.rpe,
                    bodyWeightKg: session.bodyWeightKg,
                    estimatedCaloriesBurned: session.estimatedCaloriesBurned,
                    note: session.note
                )
            },

            weightLogs: weightLogs.map { log in
                WeightLogDTO(date: log.date, weightKg: log.weightKg, note: log.note)
            },

            measurementLogs: measurementLogs.map { log in
                BodyMeasurementLogDTO(
                    date: log.date,
                    waistCm: log.waistCm,
                    chestCm: log.chestCm,
                    hipsCm: log.hipsCm,
                    armCm: log.armCm,
                    thighCm: log.thighCm
                )
            },

            dayStatuses: dayStatuses.map { status in
                DayStatusDTO(date: status.date, type: status.type.rawValue)
            },

            snapshots: snapshots.map { snapshot in
                DailyTargetSnapshotDTO(
                    date: snapshot.date,
                    goalMode: snapshot.goalMode.rawValue,
                    goalPace: snapshot.goalPace.rawValue,
                    calories: snapshot.calories,
                    proteinGrams: snapshot.proteinGrams,
                    carbsGrams: snapshot.carbsGrams,
                    fatGrams: snapshot.fatGrams,
                    fiberGrams: snapshot.fiberGrams,
                    trainingCalories: snapshot.trainingCalories
                )
            },

            favorites: favorites.map { favorite in
                FavoriteFoodDTO(
                    name: favorite.name,
                    grams: favorite.grams,
                    calories: favorite.calories,
                    proteinGrams: favorite.proteinGrams,
                    carbsGrams: favorite.carbsGrams,
                    fatGrams: favorite.fatGrams,
                    fiberGrams: favorite.fiberGrams
                )
            },

            savedMeals: savedMeals.map { meal in
                SavedMealDTO(
                    name: meal.name,
                    createdAt: meal.createdAt,
                    items: meal.items.map { item in
                        MealItemDTO(
                            name: item.name,
                            grams: item.grams,
                            calories: item.calories,
                            proteinGrams: item.proteinGrams,
                            carbsGrams: item.carbsGrams,
                            fatGrams: item.fatGrams,
                            fiberGrams: item.fiberGrams
                        )
                    }
                )
            },

            foodProducts: foodProducts.map { product in
                FoodProductDTO(
                    name: product.name,
                    brand: product.brand,
                    barcode: product.barcode,
                    caloriesPer100g: product.caloriesPer100g,
                    proteinPer100g: product.proteinPer100g,
                    carbsPer100g: product.carbsPer100g,
                    fatPer100g: product.fatPer100g,
                    fiberPer100g: product.fiberPer100g,
                    createdAt: product.createdAt
                )
            },

            mealTemplates: mealTemplates.map { template in
                MealTemplateDTO(
                    name: template.name,
                    category: template.category.rawValue,
                    calories: template.calories,
                    proteinGrams: template.proteinGrams,
                    carbsGrams: template.carbsGrams,
                    fatGrams: template.fatGrams,
                    fiberGrams: template.fiberGrams,
                    createdAt: template.createdAt
                )
            }
        )
    }

    static func encode(_ payload: BackupPayload) throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(payload)
    }

    static func decode(_ data: Data) throws -> BackupPayload {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(BackupPayload.self, from: data)
    }

    // MARK: - Restore (destructief — vervangt alle huidige data, geen samenvoegen)

    static func restore(_ payload: BackupPayload, into context: ModelContext) throws {

        try context.delete(model: UserProfile.self)
        try context.delete(model: GoalPeriod.self)
        try context.delete(model: FoodLogEntry.self)
        try context.delete(model: TrainingSession.self)
        try context.delete(model: WeightLog.self)
        try context.delete(model: BodyMeasurementLog.self)
        try context.delete(model: DayStatus.self)
        try context.delete(model: DailyTargetSnapshot.self)
        try context.delete(model: FavoriteFood.self)
        try context.delete(model: SavedMeal.self)
        try context.delete(model: MealItem.self)
        try context.delete(model: FoodProduct.self)
        try context.delete(model: MealTemplate.self)

        if let profileDTO = payload.profile {
            let profile = UserProfile(
                name: profileDTO.name,
                age: profileDTO.age,
                sex: Sex(rawValue: profileDTO.sex) ?? .male,
                heightCm: profileDTO.heightCm,
                currentWeightKg: profileDTO.currentWeightKg,
                estimatedBodyFatPercentage: profileDTO.estimatedBodyFatPercentage,
                goalMode: GoalMode(rawValue: profileDTO.goalMode) ?? .maintenance,
                goalPace: GoalPace(rawValue: profileDTO.goalPace) ?? .normal,
                activityLevel: ActivityLevel(rawValue: profileDTO.activityLevel) ?? .moderate
            )
            context.insert(profile)

            for periodDTO in payload.goalPeriods {
                let period = GoalPeriod(
                    startDate: periodDTO.startDate,
                    durationWeeks: periodDTO.durationWeeks,
                    goalMode: GoalMode(rawValue: periodDTO.goalMode) ?? .maintenance,
                    goalPace: GoalPace(rawValue: periodDTO.goalPace) ?? .normal,
                    isActive: periodDTO.isActive
                )
                period.completedAt = periodDTO.completedAt
                period.calorieAdjustment = periodDTO.calorieAdjustment
                period.lastCheckInDate = periodDTO.lastCheckInDate
                period.profile = profile
                context.insert(period)
            }
        }

        for dto in payload.foodEntries {
            context.insert(FoodLogEntry(
                date: dto.date,
                mealCategory: MealCategory(rawValue: dto.mealCategory) ?? .other,
                name: dto.name,
                grams: dto.grams,
                calories: dto.calories,
                proteinGrams: dto.proteinGrams,
                carbsGrams: dto.carbsGrams,
                fatGrams: dto.fatGrams,
                fiberGrams: dto.fiberGrams,
                note: dto.note
            ))
        }

        for dto in payload.trainings {
            context.insert(TrainingSession(
                date: dto.date,
                type: TrainingType(rawValue: dto.type) ?? .other,
                durationMinutes: dto.durationMinutes,
                rpe: dto.rpe,
                bodyWeightKg: dto.bodyWeightKg,
                estimatedCaloriesBurned: dto.estimatedCaloriesBurned,
                note: dto.note
            ))
        }

        for dto in payload.weightLogs {
            context.insert(WeightLog(date: dto.date, weightKg: dto.weightKg, note: dto.note))
        }

        for dto in payload.measurementLogs {
            context.insert(BodyMeasurementLog(
                date: dto.date,
                waistCm: dto.waistCm,
                chestCm: dto.chestCm,
                hipsCm: dto.hipsCm,
                armCm: dto.armCm,
                thighCm: dto.thighCm
            ))
        }

        for dto in payload.dayStatuses {
            context.insert(DayStatus(date: dto.date, type: DayStatusType(rawValue: dto.type) ?? .restDay))
        }

        for dto in payload.snapshots {
            context.insert(DailyTargetSnapshot(
                date: dto.date,
                goalMode: GoalMode(rawValue: dto.goalMode) ?? .maintenance,
                goalPace: GoalPace(rawValue: dto.goalPace) ?? .normal,
                calories: dto.calories,
                proteinGrams: dto.proteinGrams,
                carbsGrams: dto.carbsGrams,
                fatGrams: dto.fatGrams,
                fiberGrams: dto.fiberGrams,
                trainingCalories: dto.trainingCalories
            ))
        }

        for dto in payload.favorites {
            context.insert(FavoriteFood(
                name: dto.name,
                grams: dto.grams,
                calories: dto.calories,
                proteinGrams: dto.proteinGrams,
                carbsGrams: dto.carbsGrams,
                fatGrams: dto.fatGrams,
                fiberGrams: dto.fiberGrams
            ))
        }

        for mealDTO in payload.savedMeals {
            let meal = SavedMeal(name: mealDTO.name, createdAt: mealDTO.createdAt)
            context.insert(meal)

            for itemDTO in mealDTO.items {
                let item = MealItem(
                    name: itemDTO.name,
                    grams: itemDTO.grams,
                    calories: itemDTO.calories,
                    proteinGrams: itemDTO.proteinGrams,
                    carbsGrams: itemDTO.carbsGrams,
                    fatGrams: itemDTO.fatGrams,
                    fiberGrams: itemDTO.fiberGrams
                )
                meal.items.append(item)
            }
        }

        for dto in payload.foodProducts {
            context.insert(FoodProduct(
                name: dto.name,
                brand: dto.brand,
                barcode: dto.barcode,
                caloriesPer100g: dto.caloriesPer100g,
                proteinPer100g: dto.proteinPer100g,
                carbsPer100g: dto.carbsPer100g,
                fatPer100g: dto.fatPer100g,
                fiberPer100g: dto.fiberPer100g
            ))
        }

        for dto in payload.mealTemplates {
            context.insert(MealTemplate(
                name: dto.name,
                category: MealCategory(rawValue: dto.category) ?? .other,
                calories: dto.calories,
                proteinGrams: dto.proteinGrams,
                carbsGrams: dto.carbsGrams,
                fatGrams: dto.fatGrams,
                fiberGrams: dto.fiberGrams
            ))
        }

        try context.save()
    }

}

File: android_app/lib/logic/backup_manager.dart
import 'dart:convert';

import 'package:drift/drift.dart';

import '../data/database.dart';

/// Poort van `BackupManager.swift` + `BackupModels.swift`: exporteert alle data
/// naar één JSON-bestand en zet een eerdere back-up terug (vervangt alles).
/// Datums als ISO8601, enums als naam. Back-ups zijn bedoeld binnen Android
/// (net als iOS z'n lokale data) — geen cross-platform-garantie.
class BackupManager {
  BackupManager._();

  static String _iso(DateTime d) => d.toIso8601String();
  static DateTime _date(dynamic s) => DateTime.parse(s as String);

  static T _enumByName<T extends Enum>(List<T> values, dynamic name, T fallback) {
    for (final v in values) {
      if (v.name == name) return v;
    }
    return fallback;
  }

  // MARK: - Export

  static Future<String> buildJson(AppDatabase db) async {
    final profile = await db.select(db.userProfiles).getSingleOrNull();
    final goalPeriods = await db.select(db.goalPeriods).get();
    final foodEntries = await db.select(db.foodLogEntries).get();
    final trainings = await db.select(db.trainingSessions).get();
    final weightLogs = await db.select(db.weightLogs).get();
    final measurementLogs = await db.select(db.bodyMeasurementLogs).get();
    final dayStatuses = await db.select(db.dayStatuses).get();
    final snapshots = await db.select(db.dailyTargetSnapshots).get();
    final favorites = await db.select(db.favoriteFoods).get();
    final savedMeals = await db.select(db.savedMeals).get();
    final mealItems = await db.select(db.mealItems).get();
    final foodProducts = await db.select(db.foodProducts).get();
    final mealTemplates = await db.select(db.mealTemplates).get();

    final payload = <String, dynamic>{
      'exportedAt': _iso(DateTime.now()),
      'profile': profile == null
          ? null
          : {
              'name': profile.name,
              'age': profile.age,
              'sex': profile.sex.name,
              'heightCm': profile.heightCm,
              'currentWeightKg': profile.currentWeightKg,
              'targetWeightKg': profile.targetWeightKg,
              'estimatedBodyFatPercentage': profile.estimatedBodyFatPercentage,
              'goalMode': profile.goalMode.name,
              'goalPace': profile.goalPace.name,
              'activityLevel': profile.activityLevel.name,
              'createdAt': _iso(profile.createdAt),
            },
      'goalPeriods': [
        for (final p in goalPeriods)
          {
            'startDate': _iso(p.startDate),
            'durationWeeks': p.durationWeeks,
            'goalMode': p.goalMode.name,
            'goalPace': p.goalPace.name,
            'isActive': p.isActive,
            'completedAt': p.completedAt == null ? null : _iso(p.completedAt!),
            'calorieAdjustment': p.calorieAdjustment,
            'lastCheckInDate': p.lastCheckInDate == null ? null : _iso(p.lastCheckInDate!),
          }
      ],
      'foodEntries': [
        for (final e in foodEntries)
          {
            'date': _iso(e.date),
            'mealCategory': e.mealCategory.name,
            'name': e.name,
            'grams': e.grams,
            'calories': e.calories,
            'proteinGrams': e.proteinGrams,
            'carbsGrams': e.carbsGrams,
            'fatGrams': e.fatGrams,
            'fiberGrams': e.fiberGrams,
            'note': e.note,
          }
      ],
      'trainings': [
        for (final t in trainings)
          {
            'date': _iso(t.date),
            'type': t.type.name,
            'durationMinutes': t.durationMinutes,
            'rpe': t.rpe,
            'bodyWeightKg': t.bodyWeightKg,
            'estimatedCaloriesBurned': t.estimatedCaloriesBurned,
            'note': t.note,
          }
      ],
      'weightLogs': [
        for (final w in weightLogs) {'date': _iso(w.date), 'weightKg': w.weightKg, 'note': w.note}
      ],
      'measurementLogs': [
        for (final m in measurementLogs)
          {'date': _iso(m.date), 'waistCm': m.waistCm, 'chestCm': m.chestCm, 'hipsCm': m.hipsCm, 'armCm': m.armCm, 'thighCm': m.thighCm}
      ],
      'dayStatuses': [
        for (final s in dayStatuses) {'date': _iso(s.date), 'type': s.type.name}
      ],
      'snapshots': [
        for (final s in snapshots)
          {
            'date': _iso(s.date),
            'goalMode': s.goalMode.name,
            'goalPace': s.goalPace.name,
            'calories': s.calories,
            'proteinGrams': s.proteinGrams,
            'carbsGrams': s.carbsGrams,
            'fatGrams': s.fatGrams,
            'fiberGrams': s.fiberGrams,
            'trainingCalories': s.trainingCalories,
          }
      ],
      'favorites': [
        for (final f in favorites)
          {'name': f.name, 'grams': f.grams, 'calories': f.calories, 'proteinGrams': f.proteinGrams, 'carbsGrams': f.carbsGrams, 'fatGrams': f.fatGrams, 'fiberGrams': f.fiberGrams}
      ],
      'savedMeals': [
        for (final meal in savedMeals)
          {
            'name': meal.name,
            'createdAt': _iso(meal.createdAt),
            'items': [
              for (final item in mealItems.where((i) => i.savedMealId == meal.id))
                {'name': item.name, 'grams': item.grams, 'calories': item.calories, 'proteinGrams': item.proteinGrams, 'carbsGrams': item.carbsGrams, 'fatGrams': item.fatGrams, 'fiberGrams': item.fiberGrams}
            ],
          }
      ],
      'foodProducts': [
        for (final p in foodProducts)
          {
            'name': p.name,
            'brand': p.brand,
            'barcode': p.barcode,
            'caloriesPer100g': p.caloriesPer100g,
            'proteinPer100g': p.proteinPer100g,
            'carbsPer100g': p.carbsPer100g,
            'fatPer100g': p.fatPer100g,
            'fiberPer100g': p.fiberPer100g,
            'createdAt': _iso(p.createdAt),
          }
      ],
      'mealTemplates': [
        for (final t in mealTemplates)
          {
            'name': t.name,
            'category': t.category.name,
            'calories': t.calories,
            'proteinGrams': t.proteinGrams,
            'carbsGrams': t.carbsGrams,
            'fatGrams': t.fatGrams,
            'fiberGrams': t.fiberGrams,
            'createdAt': _iso(t.createdAt),
          }
      ],
    };

    return const JsonEncoder.withIndent('  ').convert(payload);
  }

  // MARK: - Import (vervangt alles)

  static Future<void> restore(AppDatabase db, String jsonString) async {
    final data = jsonDecode(jsonString) as Map<String, dynamic>;

    await db.transaction(() async {
      // Alles wissen (mealItems vóór savedMeals i.v.m. de foreign key).
      await db.delete(db.mealItems).go();
      await db.delete(db.savedMeals).go();
      await db.delete(db.goalPeriods).go();
      await db.delete(db.userProfiles).go();
      await db.delete(db.foodLogEntries).go();
      await db.delete(db.trainingSessions).go();
      await db.delete(db.weightLogs).go();
      await db.delete(db.bodyMeasurementLogs).go();
      await db.delete(db.dayStatuses).go();
      await db.delete(db.dailyTargetSnapshots).go();
      await db.delete(db.favoriteFoods).go();
      await db.delete(db.foodProducts).go();
      await db.delete(db.mealTemplates).go();

      int? profileId;
      final profile = data['profile'] as Map<String, dynamic>?;
      if (profile != null) {
        profileId = await db.into(db.userProfiles).insert(UserProfilesCompanion.insert(
          name: profile['name'] as String,
          age: profile['age'] as int,
          sex: _enumByName(Sex.values, profile['sex'], Sex.male),
          heightCm: (profile['heightCm'] as num).toDouble(),
          currentWeightKg: (profile['currentWeightKg'] as num).toDouble(),
          targetWeightKg: Value((profile['targetWeightKg'] as num?)?.toDouble()),
          estimatedBodyFatPercentage: Value((profile['estimatedBodyFatPercentage'] as num?)?.toDouble()),
          goalMode: _enumByName(GoalMode.values, profile['goalMode'], GoalMode.maintenance),
          goalPace: _enumByName(GoalPace.values, profile['goalPace'], GoalPace.normal),
          activityLevel: _enumByName(ActivityLevel.values, profile['activityLevel'], ActivityLevel.moderate),
          createdAt: _date(profile['createdAt']),
        ));
      }

      for (final p in (data['goalPeriods'] as List? ?? [])) {
        await db.into(db.goalPeriods).insert(GoalPeriodsCompanion.insert(
          profileId: Value(profileId),
          startDate: _date(p['startDate']),
          durationWeeks: p['durationWeeks'] as int,
          goalMode: _enumByName(GoalMode.values, p['goalMode'], GoalMode.maintenance),
          goalPace: _enumByName(GoalPace.values, p['goalPace'], GoalPace.normal),
          isActive: Value(p['isActive'] as bool? ?? false),
          completedAt: Value(p['completedAt'] == null ? null : _date(p['completedAt'])),
          calorieAdjustment: Value((p['calorieAdjustment'] as num?)?.toDouble() ?? 0),
          lastCheckInDate: Value(p['lastCheckInDate'] == null ? null : _date(p['lastCheckInDate'])),
        ));
      }

      for (final e in (data['foodEntries'] as List? ?? [])) {
        await db.into(db.foodLogEntries).insert(FoodLogEntriesCompanion.insert(
          date: _date(e['date']),
          mealCategory: _enumByName(MealCategory.values, e['mealCategory'], MealCategory.other),
          name: e['name'] as String,
          grams: (e['grams'] as num).toDouble(),
          calories: (e['calories'] as num).toDouble(),
          proteinGrams: (e['proteinGrams'] as num).toDouble(),
          carbsGrams: (e['carbsGrams'] as num).toDouble(),
          fatGrams: (e['fatGrams'] as num).toDouble(),
          fiberGrams: (e['fiberGrams'] as num).toDouble(),
          note: Value(e['note'] as String?),
        ));
      }

      for (final t in (data['trainings'] as List? ?? [])) {
        await db.into(db.trainingSessions).insert(TrainingSessionsCompanion.insert(
          date: _date(t['date']),
          type: _enumByName(TrainingType.values, t['type'], TrainingType.other),
          durationMinutes: t['durationMinutes'] as int,
          rpe: t['rpe'] as int,
          bodyWeightKg: (t['bodyWeightKg'] as num).toDouble(),
          estimatedCaloriesBurned: (t['estimatedCaloriesBurned'] as num).toDouble(),
          note: Value(t['note'] as String?),
        ));
      }

      for (final w in (data['weightLogs'] as List? ?? [])) {
        await db.into(db.weightLogs).insert(WeightLogsCompanion.insert(
          date: _date(w['date']),
          weightKg: (w['weightKg'] as num).toDouble(),
          note: Value(w['note'] as String?),
        ));
      }

      for (final m in (data['measurementLogs'] as List? ?? [])) {
        await db.into(db.bodyMeasurementLogs).insert(BodyMeasurementLogsCompanion.insert(
          date: _date(m['date']),
          waistCm: Value((m['waistCm'] as num?)?.toDouble()),
          chestCm: Value((m['chestCm'] as num?)?.toDouble()),
          hipsCm: Value((m['hipsCm'] as num?)?.toDouble()),
          armCm: Value((m['armCm'] as num?)?.toDouble()),
          thighCm: Value((m['thighCm'] as num?)?.toDouble()),
        ));
      }

      for (final s in (data['dayStatuses'] as List? ?? [])) {
        await db.into(db.dayStatuses).insert(DayStatusesCompanion.insert(
          date: _date(s['date']),
          type: _enumByName(DayStatusType.values, s['type'], DayStatusType.restDay),
        ));
      }

      for (final s in (data['snapshots'] as List? ?? [])) {
        await db.into(db.dailyTargetSnapshots).insert(DailyTargetSnapshotsCompanion.insert(
          date: _date(s['date']),
          goalMode: _enumByName(GoalMode.values, s['goalMode'], GoalMode.maintenance),
          goalPace: _enumByName(GoalPace.values, s['goalPace'], GoalPace.normal),
          calories: (s['calories'] as num).toDouble(),
          proteinGrams: (s['proteinGrams'] as num).toDouble(),
          carbsGrams: (s['carbsGrams'] as num).toDouble(),
          fatGrams: (s['fatGrams'] as num).toDouble(),
          fiberGrams: (s['fiberGrams'] as num).toDouble(),
          trainingCalories: (s['trainingCalories'] as num).toDouble(),
        ));
      }

      for (final f in (data['favorites'] as List? ?? [])) {
        await db.into(db.favoriteFoods).insert(FavoriteFoodsCompanion.insert(
          name: f['name'] as String,
          grams: (f['grams'] as num).toDouble(),
          calories: (f['calories'] as num).toDouble(),
          proteinGrams: (f['proteinGrams'] as num).toDouble(),
          carbsGrams: (f['carbsGrams'] as num).toDouble(),
          fatGrams: (f['fatGrams'] as num).toDouble(),
          fiberGrams: (f['fiberGrams'] as num).toDouble(),
        ));
      }

      for (final meal in (data['savedMeals'] as List? ?? [])) {
        final mealId = await db.into(db.savedMeals).insert(SavedMealsCompanion.insert(
          name: meal['name'] as String,
          createdAt: _date(meal['createdAt']),
        ));
        for (final item in (meal['items'] as List? ?? [])) {
          await db.into(db.mealItems).insert(MealItemsCompanion.insert(
            savedMealId: mealId,
            name: item['name'] as String,
            grams: (item['grams'] as num).toDouble(),
            calories: (item['calories'] as num).toDouble(),
            proteinGrams: (item['proteinGrams'] as num).toDouble(),
            carbsGrams: (item['carbsGrams'] as num).toDouble(),
            fatGrams: (item['fatGrams'] as num).toDouble(),
            fiberGrams: (item['fiberGrams'] as num).toDouble(),
          ));
        }
      }

      for (final p in (data['foodProducts'] as List? ?? [])) {
        await db.into(db.foodProducts).insert(FoodProductsCompanion.insert(
          name: p['name'] as String,
          brand: Value(p['brand'] as String?),
          barcode: Value(p['barcode'] as String?),
          caloriesPer100g: (p['caloriesPer100g'] as num).toDouble(),
          proteinPer100g: (p['proteinPer100g'] as num).toDouble(),
          carbsPer100g: (p['carbsPer100g'] as num).toDouble(),
          fatPer100g: (p['fatPer100g'] as num).toDouble(),
          fiberPer100g: (p['fiberPer100g'] as num).toDouble(),
          createdAt: _date(p['createdAt']),
        ));
      }

      for (final t in (data['mealTemplates'] as List? ?? [])) {
        await db.into(db.mealTemplates).insert(MealTemplatesCompanion.insert(
          name: t['name'] as String,
          category: _enumByName(MealCategory.values, t['category'], MealCategory.other),
          calories: (t['calories'] as num).toDouble(),
          proteinGrams: (t['proteinGrams'] as num).toDouble(),
          carbsGrams: (t['carbsGrams'] as num).toDouble(),
          fatGrams: (t['fatGrams'] as num).toDouble(),
          fiberGrams: (t['fiberGrams'] as num).toDouble(),
          createdAt: _date(t['createdAt']),
        ));
      }
    });
  }
}
